axum-macros/src/debug_handler.rs

```diff
@@ -1,87 +1,62 @@
 use crate::{
-    attr_parsing::{parse_assignment_attribute, second},
+    attr_parsing::second,
     with_position::{Position, WithPosition},
 };
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote, quote_spanned};
-use syn::{parse::Parse, parse_quote, spanned::Spanned, FnArg, ItemFn, Token, Type};
+use syn::{parse_quote, spanned::Spanned, FnArg, ItemFn, Type};
 
-pub(crate) fn expand(attr: Attrs, item_fn: ItemFn) -> TokenStream {
-    let Attrs { body_ty, state_ty } = attr;
+use self::attr::Attrs;
+use self::specializer::Specializer;
+
+mod attr;
+mod specializer;
 
+pub(crate) fn expand(attr: Attrs, item_fn: ItemFn) -> TokenStream {
+    let Attrs {
+        body_ty,
+        state_ty,
+        with_tys,
+    } = attr;
     let body_ty = body_ty
         .map(second)
         .unwrap_or_else(|| parse_quote!(axum::body::Body));
-
     let mut state_ty = state_ty.map(second);
 
+    // these checks don't require the specializer, so we can generate code for them regardless
+    // of whether we can successfully create one
     let check_extractor_count = check_extractor_count(&item_fn);
     let check_path_extractor = check_path_extractor(&item_fn);
-    let check_output_impls_into_response = check_output_impls_into_response(&item_fn);
-
-    // If the function is generic, we can't reliably check its inputs or whether the future it
-    // returns is `Send`. Skip those checks to avoid unhelpful additional compiler errors.
-    let check_inputs_and_future_send = if item_fn.sig.generics.params.is_empty() {
-        if state_ty.is_none() {
-            state_ty = state_type_from_args(&item_fn);
-        }
-
-        let state_ty = state_ty.unwrap_or_else(|| syn::parse_quote!(()));
-
-        let check_inputs_impls_from_request =
-            check_inputs_impls_from_request(&item_fn, &body_ty, state_ty);
-        let check_future_send = check_future_send(&item_fn);
 
-        quote! {
-            #check_inputs_impls_from_request
-            #check_future_send
+    if state_ty.is_none() {
+        state_ty = state_type_from_args(&item_fn);
+    }
+    let state_ty = state_ty.unwrap_or_else(|| syn::parse_quote!(()));
+
+    // If the function is generic and and improper `with` statement was provided to the macro, we can't
+    // reliably check its inputs or outputs. This will result in an error. We skip those checks to avoid
+    // unhelpful additional compiler errors.
+    let specializer_checks = match Specializer::new(with_tys, &item_fn) {
+        Ok(specializer) => {
+            let check_output_impls_into_response =
+                check_output_impls_into_response(&item_fn, &specializer);
+            let check_inputs_impls_from_request =
+                check_inputs_impls_from_request(&item_fn, &body_ty, state_ty, &specializer);
+            let check_future_send = check_future_send(&item_fn, &specializer);
```

Comment:
I think it would be neat if these function were methods on `Specializer`.

Reply:
Also it seems unnecessary to pass the `item_fn` to `Specializer` and then to all these functions separate that have access to the `Specializer`.